tagmedia

```diff
@@ -30,127 +48,481 @@ GetOptions(
   }
 
-  $digest = $digest->hexdigest;
+  die "tags too large: \"$buf\"\n" if length($buf) > 0x200;
+  $buf .= " " x (0x200 - length($buf));
+
+  my $len;
+  die "$image->{name}: $!\n" unless open $image->{fh}, "+<", $image->{name};
+  die "$image->{name}: $!\n" unless seek $image->{fh}, 0x8373, 0;
+  die "$image->{name}: $!\n" unless $len = syswrite $image->{fh}, $buf, 0x200;
+  die "$image->{name}: failed to update image\n" unless $len == 0x200;
 
-  unshift @opt_add_tag, "${opt_digest}sum=$digest";
+  close $image->{fh};
 }
 
-close F;
 
-# replace existing tags
-for (@opt_add_tag) {
-  $new_tag{$1} = 1 if /^(\S+?)\s*=/;
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# tag = parse_tag(string)
+#
+# Parse line ('key=value' format) and return hash ref with key/value elements.
+# value is undef if there's no '='.
+#
+sub parse_tag
+{
+  my ($line) = @_;
+
+  $line =~ s/^\s*|\s*$//g;
+
+  if($line =~ /^([A-Z_\d\s]+?)\s*=\s*+(.*)$/i) {
+    return { key => $1, value => $2 };
+  }
+  else {
+    return { key => $line };
+  }
 }
 
-# use old tags unless 'clean' option was given
-if(!$opt_clean) {
-  for (split /;/, $tag) {
-    s/^\s*|\s*$//g;
-    push @old_tags, "$_" unless $_ eq "" || (/^(\S+?)\s*=/ && $new_tag{$1});
+
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# tag = get_tag(tags, key)
+#
+# Get tag from tag array with specified key.
+# Returns undef if no tag with such a key exists.
+#
+# tags: array ref with tag hashes
+# key: string
+# tag: tag hash ref ('key', 'value' pair)
+#
+sub get_tag
+{
+  my ($tags, $key) = @_;
+
+  for my $tag (@$tags) {
+    return $tag if $tag->{key} eq $key;
   }
+
+  return undef;
 }
 
-# remove tags
-$rtags{$_} = 1 for @opt_remove_tag;
-for (@old_tags, @opt_add_tag) {
-  push @tags, $_ unless $_ eq "" || (/^(\S+?)\s*=/ && $rtags{$1}) ;
+
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# set_tag(tags, tag)
+#
+# Set tag in tags array.
+#
+# tags: array ref with tag hashes
+# tag: tag hash ref ('key', 'value' pair)
+#
+sub set_tag
+{
+  my ($tags, $tag) = @_;
+
+  my $old_tag = get_tag $tags, $tag->{key};
+  if($old_tag) {
+    $old_tag->{value} = $tag->{value};
+  }
+  else {
+    push @$tags, $tag;
+  }
 }
 
-$new_tag = join ';', @tags;
-die "too many tags: \"$new_tag\"\n" if length($new_tag) > 0x200;
-$new_tag .= " " x (0x200 - length($new_tag));
 
-if($write_iso) {
-  die "$iso: $!\n" unless open F, "+<$iso";
-  die "$iso: $!\n" unless seek F, 0x8373, 0;
-  die "$iso: $!\n" unless $l = syswrite F, $new_tag, 0x200;
-  die "$iso: file too short\n" unless $l = 0x200;
-  close F;
+# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+# tags = remove_tag(tags, key)
+#
+# Remove tag from tag array with specified key.
+# Returns new tag array.
+#
+# tags: array ref with tag hashes
+# key: string
+#
+sub remove_tag
+{
+  my ($tags, $key) = @_;
+  my @l;
```

Comment:
this is too much perl magic for me. What this do?

Reply:
just passed the first two arguments

Reply:
well, the first line I know, common getting perl params. But what is this `my @l` I do not see anywhere used this 'l' variable below and my should limit it to method scope, not?

Reply:
Ah, sorry. Some dead code.